modules/standard/GMP.chpl

```diff
@@ -1421,179 +1597,565 @@ module GMP {
       }
       return ret.safeCast(uint);
     }
     proc size():size_t
     {
       var ret:size_t;
-      on this {
+      on Locales[this.locale_id] {
         ret = mpz_size(this.mpz);
       }
       return ret;
     }
     proc debugprint() {
       writeln("On locale ",this.locale);
-      on this {
+      on Locales[this.locale_id] {
         chpl_gmp_mpz_print(this.mpz);
       }
     }
+
+  }
+
+  // Assignment function -- deep copies `rhs` into `lhs`
+   proc =(ref lhs: BigInt, rhs: BigInt) {
+    inline proc helpMe(ref lhs: BigInt, rhs: BigInt) {
+      if _local || rhs.locale_id == lhs.locale_id {
+        lhs.reinitBigInt(rhs.mpz, needToCopy=true);
+      } else {
+        on Locales[lhs.locale_id]{
+          var mpz_struct = rhs.mpz[1];
+          chpl_gmp_get_mpz(lhs.mpz, rhs.locale_id, mpz_struct);
+          lhs.owned = true; //remote assignment makes a deep copy
+        }
+      }
+    }
+    if _local || lhs.locale_id == chpl_nodeID {
+      helpMe(lhs, rhs);
+    }
+    else {
+      on __primitive("chpl_on_locale_num", chpl_buildLocaleID(lhs.locale_id, c_sublocid_any)) {
+        helpMe(lhs, rhs);
+      }
+    }
+  }
+
+  // assignment operator overload for signed int assignment
+  proc =(ref lhs: BigInt, rhs: int) {
+    //TODO: does this need an "on Locales[lhs.locale_id]" ?
+    lhs.set(rhs);
   }
 
+  // autoCopy and initCopy can be called by the compiler
+  pragma "donor fn"
+  pragma "auto copy fn"
+  pragma "no doc"
+  proc chpl__autoCopy(ref bir: BigInt) {
+    //this pragma may not be needed
+    pragma "no auto destroy"
+    var ret : BigInt;
+    if _local || bir.locale_id == chpl_nodeID {
+      if bir.owned {
+        if ret.owned then on ret.locale do mpz_clear(ret.mpz);
+        mpz_init_set(ret.mpz, bir.mpz);
+        ret.owned = true;
+      } else {
+        ret.mpz = bir.mpz; //make a shallow copy
+        ret.owned = false;
+      }
+    } else {
+      // TODO: test this
+      var remoteMpz = bir.mpz;
+      ret.reinitBigInt(remoteMpz, needToCopy=false);
+    }
+    return ret;
+  }
+
+  pragma "init copy fn"
+  pragma "no doc"
+  proc chpl__initCopy(ref bir: BigInt) {
+    // This pragma may be unnecessary.
+    //pragma "no auto destroy"
+    var ret : BigInt;
+    if _local || bir.locale_id == chpl_nodeID {
+      if bir.owned {
+        if ret.owned then on Locales[ret.locale_id] do mpz_clear(ret.mpz);
+        mpz_init_set(ret.mpz, bir.mpz);
+        ret.owned = true;
+      } else {
+        ret.mpz = bir.mpz; // make a shalow copy
+        ret.owned = false;
+      }
+    } else {
+      ret.set(bir);
+    }
+    return ret;
+  }
 
   proc BigInt.writeThis(writer) {
-    var (acopy,a_) = this.maybeCopy();
-    var s:string = a_.get_str();
-    //gmp_asprintf(s, "%Zd", a_.mpz);
+    var s: string;
+    on Locales[this.locale_id] {
+      s = this.get_str();
+    }
     writer.write(s);
-    if acopy then delete a_;
   }
 
+  // Operator Overloads
```

Comment:
Why are all of these declared inline?  This is going to be slow operation.  Shame on us if
the call overhead matters.